orc8r/cloud/go/services/bootstrapper/servicers/registration/registration.go

```diff
@@ -0,0 +1,107 @@
+package registration
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-openapi/strfmt"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"magma/orc8r/cloud/go/orc8r"
+	"magma/orc8r/cloud/go/serdes"
+	"magma/orc8r/cloud/go/services/bootstrapper"
+	"magma/orc8r/cloud/go/services/device"
+	"magma/orc8r/cloud/go/services/orchestrator/obsidian/models"
+	"magma/orc8r/cloud/go/services/tenants"
+	"magma/orc8r/lib/go/protos"
+)
+
+type registrationServicer struct{}
+
+func NewRegistrationServicer() (protos.RegistrationServer, error) {
+	return &registrationServicer{}, nil
+}
+
+func (r *registrationServicer) Register(c context.Context, request *protos.RegisterRequest) (*protos.RegisterResponse, error) {
+	nonce, err := NonceFromToken(request.Token)
+	if err != nil {
+		return nil, err
+	}
+
+	deviceInfo, err := bootstrapper.GetGatewayDeviceInfo(context.Background(), nonce)
+	if err != nil {
+		clientErr := makeErr(fmt.Sprintf("could not get device info from token %v: %v", request.Token, err))
+		return clientErr, nil
+	}
+
+	err = RegisterDevice(*deviceInfo, request.Hwid, request.ChallengeKey)
+	if err != nil {
+		clientErr := makeErr(fmt.Sprintf("error registering device: %v", err))
+		return clientErr, nil
+	}
+
+	controlProxy, err := GetControlProxy(deviceInfo.NetworkId)
+	if err != nil {
+		clientErr := makeErr(fmt.Sprintf("error getting control proxy: %v", err))
+		return clientErr, nil
+	}
+
+	res := &protos.RegisterResponse{
+		Response: &protos.RegisterResponse_ControlProxy{
+			ControlProxy: controlProxy,
+		},
+	}
+	return res, nil
+}
+
+var RegisterDevice = func(deviceInfo protos.GatewayDeviceInfo, hwid *protos.AccessGatewayID, challengeKey *protos.ChallengeKey) error {
```

Comment:
Earlier I said it made sense to mock out these two functions - `RegisterDevice`, and `GetControlProxy`, but after squinting at this PR a little more, I would suggest a minor change.

Keep these two functions as private methods, ie. `registerDevice` and `getControlProxy`, and instead of mocking them out, mock out `device.RegisterDevice`, `tenants.GetAllTenants`, and set an expectation on `tenants.GetControlProxy` getting called with the expected `tenantID`.

`RegisterDevice` and `GetControlProxy` are basically private methods, so would be at the wrong layer for mocking. Mock and set expectations on public facing methods instead.

Reply:
I think Hunter also mentioned that there was another way to mock out without doing it like this. Do you think the current mocking + generating mocks would be good?

> Here’s how we encapsulate a mock into non-global state: https://github.com/magma/magma/blob/1dffd37504977682fc185e28dee7bde592b9717d/orc8r/cloud/go/services/state/indexer/indexer.go#L14

Ref: https://magmacore.slack.com/archives/C02GNL7EG9E/p1638818217000300?thread_ts=1638568745.000400&cid=C02GNL7EG9E

Reply:
Also, I don't think I can do what you said? Like I tried to, but I can't set them unless I make them (EG `device.RegisterDevice`) into global states (`var xxxx = func(...) (...) { ...}`) which we might not want to do for such public endpoints? (since lots of services can use the RPC endpoints)

Reply:
One option is to modify `registrationServicer` struct so you can pass in the dependencies, ie. the functions `device.RegisterDevice`, etc. in this example.

Note: Making something like `device.RegisterDevice` into a global is a little strange, I agree, but I would argue you're already doing a very similar thing here in this PR in `client_api.go` (I don't think this is needed if you go with the dependency injection option)

Reply:
Ok that sounds good. Do you think generating mocks would keep it clean as well? I'm thinking of trying that approach

Reply:
Just kidding, I think I'll do the dependency injection -- it looks pretty good